rai/node/bootstrap.cpp

```diff
@@ -939,29 +968,84 @@ bool rai::bootstrap_attempt::consume_future (std::future<bool> & future_a)
 	return result;
 }
 
+struct block_rate_cmp
+{
+	bool operator() (const std::shared_ptr<rai::bootstrap_client> & lhs, const std::shared_ptr<rai::bootstrap_client> & rhs) const
+	{
+		return lhs->block_rate () > rhs->block_rate ();
+	}
+};
+
 void rai::bootstrap_attempt::populate_connections ()
 {
-	if (connections < node->config.bootstrap_connections)
+	double rate_sum = 0.0;
+	std::priority_queue<std::shared_ptr<rai::bootstrap_client>, std::vector<std::shared_ptr<rai::bootstrap_client>>, block_rate_cmp> sorted_connections;
 	{
-		auto peer (node->peers.bootstrap_peer ());
-		if (peer != rai::endpoint (boost::asio::ip::address_v6::any (), 0))
+		std::unique_lock<std::mutex> lock (mutex);
+		for (auto & c : clients)
 		{
-			auto client (std::make_shared<rai::bootstrap_client> (node, shared_from_this (), rai::tcp_endpoint (peer.address (), peer.port ())));
-			client->run ();
-			std::lock_guard<std::mutex> lock (mutex);
-			clients.push_back (client);
+			if (auto client = c.lock ())
+			{
+				double elapsed = client->elapsed_seconds ();
+				auto rate = client->block_rate ();
+				rate_sum += rate;
+				if (client->elapsed_seconds () > 5.0 && client->block_count > 0)
+				{
+					sorted_connections.push (client);
+				}
+				// TODO - make configurable, hard stop.
+				if (elapsed > 30 && rate < 10)
+				{
+					client->stop (true);
+				}
+			}
 		}
-		else
+	}
+
+	if (sorted_connections.size () >= (node->config.bootstrap_connections * 2 - 1) / 3 && node->config.bootstrap_connections > 1)
```

Comment:
Can you decompose this expression so it's easier to digest, or give it a name with a local `bool` or a function that returns `bool`?